### PR TITLE
storage_proxy: use small_vector for vectors of inet_address

### PR DESCRIPTION
**Pull request by avikivity**
storage_proxy uses std::vector<inet_address> for small lists of nodes - for replication (often 2-3 replicas per operation) and for pending operations (usually 0-1). These vectors require an allocation, sometimes more than one if reserve() is not used correctly.

This series switches storage_proxy to use utils::small_vector instead, removing the allocations in the common case.

Test results (perf_simple_query --smp 1 --task-quota-ms 10):

```
before: median 184810.98 tps ( 91.1 allocs/op,  20.1 tasks/op,   54564 insns/op)
after:  median 192125.99 tps ( 87.1 allocs/op,  20.1 tasks/op,   53673 insns/op)
```

4 allocations and ~900 instructions are removed (the tps figure is also improved, but it is less reliable due to cpu frequency changes).

The type change is unfortunately not contained in storage_proxy - the abstraction leaks to providers of replica sets and topology change vectors. This is sad but IMO the benefits make it worthwhile.

I expect more such changes can be applied in storage_proxy, specifically std::unordered_set<gms::inet_address> and vectors of response handles.

### CODE REVIEW

**Comment by tgrabiec** [comment]
184k tps seems like a very high score. Where did you run it?

File: db/hints/manager.hh [hunk 1]
```diff
@@ -35,7 +35,7 @@
 #include <seastar/core/expiring_fifo.hh>
 #include "lister.hh"
 #include "gms/gossiper.hh"
-#include "locator/snitch_base.hh"
+#include "inet_address_vectors.hh"
```
Comment (gleb-cloudius) [inline db/hints/manager.hh:38]: Why did you replace include instead of adding one?

File: inet_address_vectors.hh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+#include "gms/inet_address.hh"
+#include <vector>
+
+using inet_address_vector_replica_set = std::vector<gms::inet_address>;
+
+using inet_address_vector_topology_change = std::vector<gms::inet_address>;
```
Comment (gleb-cloudius) [inline inet_address_vectors.hh]: I fail to see the need for two types. What is special about the second one?

Comment (gleb-cloudius) [reply inet_address_vectors.hh]: Also why put inet_address_vector in there if you want to hide exactly those details behind a typedef? Later we may replace inet_address with host id (uuid) weill we do another rename?

Comment (psarna) [reply inet_address_vectors.hh]: The mystery will be unveiled in the next patch - these types are identical now, but will differ in the space reserved for a small vector.

Comment (gleb-cloudius) [reply inet_address_vectors.hh]: So the reason for all the hoopla is to save 16 bytes on stack? And we want to allow arbitrary config changes anyway, so topology change will have more then one element.

File: repair/repair.cc [hunk 3]
```diff
@@ -284,7 +284,7 @@ static std::vector<gms::inet_address> get_neighbors(database& db,
         ret.erase(it, ret.end());
     }
 
-    return ret;
+    return boost::copy_range<std::vector<gms::inet_address>>(std::move(ret));
```
Comment (gleb-cloudius) [inline repair/repair.cc:287]: Why is this needed?

File: db/hints/manager.hh (hunk 1 above)
Comment (avikivity) [reply db/hints/manager.hh:38]: This might have been accidental.

File: inet_address_vectors.hh (hunk 2 above)
Comment (avikivity) [reply inet_address_vectors.hh]: It's more than 16 bytes, since inet_address supports ipv6.

Comment (avikivity) [reply inet_address_vectors.hh]: It allows any size vector, this is just how much we optimize for.

**Review by tgrabiec — APPROVED** [review]

**Comment by avikivity** [comment]
> 184k tps seems like a very high score. Where did you run it?

```
Core(s) per socket:              16
Model name:                      AMD Ryzen 9 5950X 16-Core Processor
```

Highly recommended.

**Comment by avikivity** [comment]
v2:
 - drop unrelated removal of some random include from hints/manager.hh
 - add copyright/license to new header

**Review by gleb-cloudius — APPROVED** [review]
